workers/src/executor_def.rs

```diff
@@ -0,0 +1,152 @@
+use crate::abi::L0V2EndpointAbi::PacketSent;
+use crate::abi::L0V2EndpointAbi::PacketVerified;
+use crate::abi::SendLibraryAbi::ExecutorFeePaid;
+use crate::chain::connections::build_executor_subscriptions;
+use crate::chain::connections::get_abi_from_path;
+use crate::chain::connections::get_http_provider;
+use crate::chain::contracts::create_contract_instance;
+use crate::chain::contracts::{lz_receive, prepare_header};
+use crate::chain::ContractInst;
+use crate::config::DVNConfig;
+use alloy::dyn_abi::DynSolValue;
+use alloy::primitives::I256;
+use eyre::Result;
+use futures::StreamExt;
+use std::collections::VecDeque;
+use std::time::Duration;
+use tokio::time::sleep;
+use tracing::error;
+
+pub struct NFFLExecutor {
+    config: DVNConfig,
+    packet_queue: VecDeque<PacketSent>,
+    finish: bool,
+}
+
+impl NFFLExecutor {
+    pub(crate) const MAX_RETRIES: usize = 10;
+    pub(crate) const NOT_EXECUTABLE: &'static DynSolValue = &DynSolValue::Int(I256::ZERO, 32);
+    pub(crate) const VERIFIED_NOT_EXECUTABLE: &'static DynSolValue = &DynSolValue::Int(I256::ONE, 32);
+
+    pub fn new(config: DVNConfig) -> Self {
+        NFFLExecutor {
+            config,
+            packet_queue: VecDeque::new(),
+            finish: false,
+        }
+    }
+
+    pub fn finish(&mut self) {
+        // Note: we are in a single-threaded event loop, and we do not care about atomicity (yet).
+        self.finish = true;
+        self.packet_queue.clear();
+    }
+
+    pub async fn listen(&mut self) -> Result<()> {
+        let (mut ps_stream, mut ef_stream, mut pv_stream) = build_executor_subscriptions(&self.config).await?;
+
+        let http_provider = get_http_provider(&self.config)?;
+        let l0_abi = get_abi_from_path("./abi/L0V2Endpoint.json")?;
+        // Create a contract instance.
+        let contract = create_contract_instance(self.config.l0_endpoint_addr, http_provider, l0_abi)?;
+
+        loop {
+            tokio::select! {
+                Some(log) = ps_stream.next() => {
+                    match log.log_decode::<PacketSent>() {
+                        Ok(packet_log) => {
+                            self.packet_queue.push_back(packet_log.data().clone());
+                        },
+                        Err(e) => { error!("Failed to decode PacketSent event: {:?}", e);}
+                    }
+                }
+                Some(log) = ef_stream.next() => {
+                    match log.log_decode::<ExecutorFeePaid>() {
+                        Ok(executor_fee_log) => {
+                            if self.packet_queue.is_empty() {
+                                continue;
+                            }
+
+                            if !executor_fee_log.data().executor.eq(&self.config.dvn_addr)  {
+                                self.packet_queue.clear();
```

Comment:
Wouldn't this cause an issue when this is multi-threaded? Maybe we can add a comment? 
it might be more appropriate to pop the packet and not clear?

Reply:
The question is -- will it be multi-threaded? If yes, we can easily move `packet_queue` to the `listen` method. Maybe I will do it now.